### PR TITLE
Remove IdeResponse, integrate it into IdeM as FreeT IdeDefer

### PR DESCRIPTION
**Pull request by Bubba**
This removes the whole `IdeResponse`/`pluginGetFileResponse` headache. Use `IdeResult` now instead.
It also removes the versions of `withCachedModule` that didn't take a default. Now they must provide a default fallback if there is no module (which most of them did anyway), removing the `NoModuleAvailable` error code.
The dispatcher now also handles all the logic for deferred responses now, whereas previously half of it was in`ModuleCache.hs` as well.

### CODE REVIEW

File: hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs [hunk 1]
```diff
@@ -198,22 +199,46 @@ instance ToJSON IdePlugins where
...
+type IdeGhcM = GM.GhcModT IdeBase
 
 instance MonadMTState IdeState IdeGhcM where
   readMTS = lift $ lift $ lift readMTS
-  modifyMTS f = lift $ lift $ lift $ modifyMTS f
+  modifyMTS = lift . lift . lift . modifyMTS
+
+-- | A computation that is deferred until the module is cached.
+-- Note that the module may not typecheck, in which case 'UriCacheFailed' is passed
+data IdeDefer a = IdeDefer FilePath (UriCache -> a) deriving Functor
```
Comment (alanz) [inline hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs]: This is probably not the right time to do it, but at some point we should let the module cache be more lenient, in the sense that if we get anything out, we call the function.

Here anything is `ParsedSource`, `RenamedSource`, `TypeCheckedSource`.

File: src/Haskell/Ide/Engine/Dispatcher.hs [hunk 2]
```diff
@@ -81,46 +80,33 @@ mainDispatcher inChan ghcChan ideChan = forever $ do
     Left r ->
       atomically $ writeTChan ideChan r
 
-ideDispatcher :: forall void m. DispatcherEnv -> ErrorHandler -> CallbackHandler m -> TChan (IdeRequest m) -> IdeM void
-ideDispatcher env errorHandler callbackHandler pin = forever $ do
+ideDispatcher :: forall void m. TVar IdeState -> J.ClientCapabilities -> DispatcherEnv -> ErrorHandler -> CallbackHandler m -> TChan (IdeRequest m) -> IO void
+ideDispatcher stateVar caps env errorHandler callbackHandler pin = flip runReaderT stateVar $ flip runReaderT caps $ forever $ do
```
Comment (alanz) [inline src/Haskell/Ide/Engine/Dispatcher.hs]: Long lines again, difficult to review

**Comment by alanz** [comment]
Nothing obvious jumps out at me.

**Comment by alanz** [comment]
@Gurkenglas, @Bubba 

Are we ok with this?  Should I merge?

@wz1000 any comment?

**Comment by wz1000** [comment]
@alanz Seems reasonable to me.

**Comment by alanz** [comment]
@Gurkenglas 229?

File: hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs [hunk 3]
```diff
@@ -198,22 +198,42 @@ instance ToJSON IdePlugins where
 
 -- ---------------------------------------------------------------------
 
-type IdeGhcM = GM.GhcModT IdeM
+type IdeGhcM = GM.GhcModT IdeBase
```
Comment (alanz) [inline hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs]: This definition should be grouped with the other three, just below.

**Comment by Bubba** [comment]
@Gurkenglas @alanz Instead of throwing an error/having an invariant, how would you feel about making `IdeM` `IdeDefer` and making `IdeBase` `IdeM`? This way we can be explicit in the type signatures of when and where you can defer. On the other hand, its added complexity

**Comment by alanz** [comment]
@Bubba I do not have strong feelings either way, except I prefer simplicity in general.

And we need to merge this thing some time.

File: hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs (hunk 1 above)
Comment (Bubba) [reply hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs]: I wrote this review 5 days ago, but didn't realise you had to submit it for you to see it...

**Pull request merged**